addOns/spider/src/main/java/org/zaproxy/zap/extension/spider/SvgHrefSpider.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2022 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.spider;
+
+import java.io.ByteArrayInputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.regex.Pattern;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import net.htmlparser.jericho.Source;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.network.HttpMessage;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXParseException;
+import org.zaproxy.zap.spider.parser.SpiderParser;
+import org.zaproxy.zap.utils.XmlUtils;
+
+public class SvgHrefSpider extends SpiderParser {
+    private static final Logger LOGGER = LogManager.getLogger(SvgHrefSpider.class);
+    private static final Pattern PATTERN_SVG_EXTENSION =
+            Pattern.compile("\\.svg\\z", Pattern.CASE_INSENSITIVE);
+    private static final String HREF_EXPRESSION = "//*[@href or @HREF]";
+
+    private static XPathExpression xpathHrefExpression;
+
+    static {
+        try {
+            XPath xpath = XPathFactory.newInstance().newXPath();
+            xpathHrefExpression = xpath.compile(HREF_EXPRESSION);
+        } catch (XPathExpressionException e) {
+            LOGGER.error(e);
+        }
+    }
+
+    private static ThreadLocal<?> docBuildeIns =
+            new ThreadLocal<Object>() {
+                @Override
+                protected DocumentBuilder initialValue() {
+                    try {
+                        return XmlUtils.newXxeDisabledDocumentBuilderFactory().newDocumentBuilder();
+                    } catch (ParserConfigurationException e) {
+                        String msg = "DocumentBuilder object initialization failed!";
+                        LOGGER.error(msg, e);
+                        throw new IllegalStateException(msg, e);
+                    }
+                }
+            };
```

Comment:
Apparently DocumentBuilder (or more specifically DocumentBuilderFactory) isn't thread friendly. I'm sure this should be renamed but I left it as-is for now so that it was obvious I went hunting for solutions and it would be easy to find the SO, Code Ranch, whatever example I had found.

I'm not set on doing it this way, but it worked and I couldn't come up with anything else.

Reply:
I'd not expect that many SVGs that it would be a contention problem to sync on the `DocumentBuilder` itself.

Reply:
I ran into the error on a local server when serving just 3 or 4 SVGs. I think it’s actually the factory that’s the issue from what I’ve read.

Reply:
Sorry I misread that, sync on the factory.

Reply:
You mean I should extract a factory var and declare it synchronized?

Reply:
https://titanwolf.org/Network/Articles/Article?AID=57ded318-8185-4bc5-87fd-2f6c9274a503
(Might need to use Reader View to get rid of the registration popup)

Reply:
Eclipse says `synchronized` is invalid for DocumentBuilderFactory either as a local var or class var 🤷

Reply:
To be clear I believe the issue is more that the spider is threaded and might be trying to parse more than one SVG at the same time. (Whereas for the sitemap.xml spider there should only be one sitemap.)

Reply:
I meant create the instance as usual and then use a synchronized block, e.g.:
```java
synchronized (documentBuilder) {
    //...
}
```